url0e7d57473a_kyorohiro_HetimaUtil_tgz-pJ8-net_hetimatan_net_http_HttpServer3xxJ8/src/net/hetimatan/net/http/HttpServerFront.java

```diff
@@ -0,0 +1,157 @@
+import net.hetimatan.util.event.net.io.KyoroSocket;
+import net.hetimatan.util.http.HttpRequestHeader;
+import net.hetimatan.util.http.LookaheadHttpBody;
+import net.hetimatan.util.http.LookaheadHttpHeader;
+import net.hetimatan.util.http.HttpRequest;
+import net.hetimatan.util.log.Log;
+
+public class HttpServerFront {
+	public static final String TAG = "HttpFront";
+
+	public String sId = ""; 
+	public long sTimeout = 10*1024;
+	private boolean mIsClose = false;
+
+	private WeakReference<HttpServer> mServer = null;
+	private KyoroSocket mSocket = null;
+	private HttpRequest mUri = null;
+	private MarkableReader mCurrentReader = null;
+	private LookaheadHttpHeader mHeaderChunk = null;
+	private KyoroFileForKyoroSocket mKFKSocket = null;
+	private LinkedList<EventTask> mMyTask = new LinkedList<EventTask>();
+
+
+	public HttpServerFront(HttpServer server, KyoroSocket socket) throws IOException {
+		sId = "[httpfront:"+socket.getHost()+ ":"+ socket.getPort()+"]";
+		socket.setDebug(sId);
+
+		mServer = new WeakReference<HttpServer>(server);
+		mSocket = socket;
+		mKFKSocket = new KyoroFileForKyoroSocket(socket, 1024);
+		mCurrentReader = new MarkableFileReader(mKFKSocket,1024);
+	}
+
+	public void addMyTask(EventTask task) {
+		mMyTask.add(task);
+	}
+
+	public boolean parseableHeader() throws IOException {
+		boolean prev = mCurrentReader.setBlockOn(false);
+		try {
+			if(mHeaderChunk == null) {
+				mHeaderChunk = new LookaheadHttpHeader(mCurrentReader, Integer.MAX_VALUE);
+			}
+			boolean parseable =LookaheadHttpHeader.readByEndOfHeader(mHeaderChunk, mCurrentReader);
+			if(parseable == true) {
+				mCurrentReader.seek(mHeaderChunk.getStart());
+				mHeaderChunk = null;
+			}
+			else if(sTimeout < mHeaderChunk.getElapsedTime()) {
+				HttpHistory.get().pushMessage(this.sId+":timeout parseheader:"+"\n");
+				mHeaderChunk = null;
+				throw new IOException("-timeout parse header-");
+			}
+			return parseable;
+		} finally {
+			mCurrentReader.setBlockOn(prev);
+		}
+	}
+
+	private LookaheadHttpBody mLookaheadBody = null;
+	public boolean parseableBody() throws IOException {
+		String contentLength = mUri.getHeaderValue(HttpRequestHeader.HEADER_CONTENT_LENGTH);
+		if(contentLength == null || contentLength.length() == 0) {return true;}
+		int _contentLength = 0;
+		try {
+			_contentLength = Integer.parseInt(contentLength);
+		} catch(Exception e) {
+			return true;
+		}
+
+		if(mLookaheadBody == null) {
+			mLookaheadBody = new LookaheadHttpBody(mCurrentReader, mCurrentReader.getFilePointer(), _contentLength); 
+		}
+
+		return mLookaheadBody.lookahead();
+	}
+
+	public void parseHeader() throws IOException {
+		this.mUri = HttpRequest.decode(mCurrentReader);
+		HttpHistory.get().pushMessage(this.sId+":parse request:"+mUri.toString()+"\n");
+	}
+
+
+	public void startResponseTask() throws IOException {
+		if(Log.ON){Log.v(TAG, "HttpFront#doRespose");}
+		HttpServerFront info = this;
+		HttpServer server = mServer.get();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *server*. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).